python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -414,49 +439,136 @@ def execute_asset_backfill_iteration(
+            # finished in order to display the final partition statuses in the UI.
+            updated_backfill = updated_backfill.with_status(BulkActionStatus.COMPLETED)
 
-    updated_backfill = backfill.with_asset_backfill_data(
-        result.backfill_data, dynamic_partitions_store=instance
-    )
-    if result.backfill_data.is_complete():
-        # The asset backfill is complete when all runs to be requested have finished (success,
+        pipeline_and_execution_plan_cache: Dict[int, Tuple[ExternalJob, ExternalExecutionPlan]] = {}
+
+        for run_request in result.run_requests:
+            yield None
+            submit_run_request(
+                run_request=run_request,
+                asset_graph=asset_graph,
+                # create a new request context for each run in case the code location server
+                # is swapped out in the middle of the backfill
+                workspace=workspace_process_context.create_request_context(),
+                instance=instance,
+                pipeline_and_execution_plan_cache=pipeline_and_execution_plan_cache,
+            )
+
+    elif backfill.status == BulkActionStatus.CANCELING:
+        # Find all cancellable runs and mark them as canceled
+        cancelable_run_ids = fetch_cancelable_run_ids_for_asset_backfill(
+            instance, backfill.backfill_id
+        )
+        if cancelable_run_ids:
+            if not instance.run_coordinator:
+                check.failed("The instance must have a run coordinator in order to cancel runs")
+            for run_id in cancelable_run_ids:
+                instance.run_coordinator.cancel_run(run_id)
+
+        # Update the asset backfill data to contain the newly materialized/failed partitions.
+        updated_asset_backfill_data = None
+        for updated_asset_backfill_data in get_canceling_asset_backfill_iteration_data(
+            backfill.backfill_id,
+            asset_backfill_data,
+            instance_queryer,
+            asset_graph,
+            backfill_start_time,
+        ):
+            yield None
+
+        if not isinstance(updated_asset_backfill_data, AssetBackfillData):
+            check.failed(
+                "Expected get_canceling_asset_backfill_iteration_data to return a PartitionBackfill"
+            )
+
+        updated_backfill = backfill.with_asset_backfill_data(
+            updated_asset_backfill_data, dynamic_partitions_store=instance
+        )
+        # The asset backfill is successfully canceled when all requested runs have finished (success,
         # failure, or cancellation). Since the AssetBackfillData object stores materialization states
         # per asset partition, the daemon continues to update the backfill data until all runs have
         # finished in order to display the final partition statuses in the UI.
-        updated_backfill = updated_backfill.with_status(BulkActionStatus.COMPLETED)
+        if updated_asset_backfill_data.have_all_requested_runs_finished():
```

Comment:
what is the ui like while we're waiting? Can I see which runs haven't finished and go force terminate them myself if necessary?